x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/writer/LengthEncodedWriter.java

```diff
@@ -27,18 +31,50 @@
  */
 public class LengthEncodedWriter implements RecordWriter {
     private OutputStream outputStream;
+    private OutputStream fileOutputStream;
     private ByteBuffer lengthBuffer;
 
+    private Logger logger = LogManager.getLogger(LengthEncodedWriter.class);
+
     /**
      * Create the writer on the OutputStream <code>os</code>.
      * This object will never close <code>os</code>.
      */
     public LengthEncodedWriter(OutputStream os) {
+        this(os, null);
+    }
+
+    public LengthEncodedWriter(OutputStream os, String filePath) {
         outputStream = os;
+        try {
+            if (filePath != null) {
+                Path file = Path.of(filePath);
+                logger.info("Opening file: " + filePath + " for writing.");
+                fileOutputStream = Files.newOutputStream(file);
+            } else {
+                fileOutputStream = null;
+            }
+        } catch (IOException e) {
+            logger.error("Failed to open file: " + filePath + " for writing.", e.getMessage(), e);
+            fileOutputStream = null;
+        }
         // This will be used to convert 32 bit integers to network byte order
         lengthBuffer = ByteBuffer.allocate(4); // 4 == sizeof(int)
     }
 
+    // Add public destructor
+    public void close() {
+        if (fileOutputStream != null) {
+            try {
+                fileOutputStream.close();
+            } catch (IOException e) {
+                logger.error("Failed to close file output stream.");
+                // log error message and stack trace
+                logger.error(e.getMessage(), e);
```

Comment:
What do you think about combining these into a single log message?

Reply:
Good point. Fixed it.